### ISSUE TITLE
Does BotMan provide a way to delay messages (e.g. to conform to API rate limits)

### ISSUE THREAD

**Issue by Coksnuss** [botman/botman#782]
- BotMan Version: 2.3.2
- PHP Version: 7.2.4
- Messaging Service(s): Telegram
- Cache Driver: FilesystemCache

### Description:
Just yesterday I found this framework and I am now considering to migrate my existing Telegram bot using this framework. My bot sometimes requires to deliver notifications to a lot of users in the least time possible. Because the Telegram API imposes a limit on how many messsages are allowed to be send at any time (1 msg/sec and 20 msg/min within a particular chat, 30 msg/sec to different chats)\[1\], the bot would very quickly start seeing 429 errors. In order to avoid hitting these limits I implemented a Guzzle middleware that transparently keeps track of the requests that are targeted to the 'sendMessage' endpoint of the Telegram API and, if necessary, delays requests to avoid hitting the limits.

**TL;DR;**
So my question is: **Is it possible to delay messages sent by the bot by BotMan (e.g. via reply()) a specified amount of time?**.

It goes without saying that I am searching for a non-blocking mechanism in order to continue code execution while delaying specific requests. However, As far as I can tell, the Interfaces that the bot uses for HTTP communication (e.g. \BotMan\BotMan\Interfaces\HttpInterface) are not supposed to work with asynchronous requests (i.e. they do not support the returning of promises). Thus, I am a bit puzzled on how to implement a throttling middleware (if, at all, possible).

\[1\](https://core.telegram.org/bots/faq#my-bot-is-hitting-limits-how-do-i-avoid-this)

**Comment by mpociot** [comment]
> In order to avoid hitting these limits I implemented a Guzzle middleware that transparently keeps track of the requests that are targeted to the 'sendMessage' endpoint of the Telegram API and, if necessary, delays requests to avoid hitting the limits.

So there's no built-in feature that you can use to enable throttling right now.
I think the most useful solution would be to create a `sending` middleware that manages the number of messages sent in a specific time interval.

**Comment by Coksnuss** [comment]
> I think the most useful solution would be to create a sending middleware

Any suggestions on implementing it in an asynchronous way (i.e. without `sleep`)

**Comment by christophrumpel** [comment]
What about using Laravel Queues. Wouldn't that be possible as well with dynamic delays?

**Comment by Coksnuss** [comment]
> What about using Laravel Queues. Wouldn't that be possible as well with dynamic delays?

Thought about this too. Problem ist that this would require the serialization of the `next` handler of the sending middleware. Consider this middleware:

```php
class ThrottleMiddleware implements Sending
{
    /**
     * @param mixed $payload
     * @param callable $next
     * @param BotMan $bot
     *
     * @return mixed
     */
    public function sending($payload, $next, BotMan $bot)
    {
        // In order to delay the sending of the message, it would be required to serialize the `$next` callback, which is not possible
        serialize($next); // error!
        return $next($payload);
    }
}
```

**Comment by christophrumpel** [comment]
Hmm I see... I just remember some time ago where I tested to use the queue system and it worked. But this was before BotMan Middlewares came. Can't remember the details. Would be a nice addition to BotMan 3.0 @mpociot to make that work?

**Comment by Coksnuss** [comment]
> Would be a nice addition to BotMan 3.0 @mpociot to make that work?

A unique selling point too, it seems

| Framework | Asynchronous support |
| -------------- | ----------------------------- |
| [php-telegram-bot/core](https://github.com/php-telegram-bot/core) | Only synchronous |
| [botstan/API](https://github.com/botstan/API) | Only synchronous |
| [botman/botman](https://github.com/botman/botman) | Only synchronous |
| [irazasyed/telegram-bot-sdk](https://github.com/irazasyed/telegram-bot-sdk) | Supports async requests but as it seems the HTTP response cannot be accessed and it is not supposed to work within an application event-loop (see [#564](https://github.com/irazasyed/telegram-bot-sdk/issues/564)) |
| [unreal4u/telegram-api](https://github.com/unreal4u/telegram-api) | Should probably work (haven't tested yet) but functions as mere API wrapper (no support for commands, conversations, ...) |

**Comment by mpociot** [comment]
I'll look back into this during the 3.0 release phase
